feat: Re-enable the Boolean control (change it from a knob to a control)

lauriharpf opened the pull request:
Issue: #177

## What I did
Enabled the Boolean control, which was commented out.

## How to test

Please explain how to test your changes and consider the following questions

- Does this need a new example in examples/native? **Yes**. Added `BooleanExample.stories.tsx` and `BooleanExampleButton.tsx` for this (see below for discussion).
- Does this need an update to the documentation? **Not sure?**

## Discussion
The 6.0 controls and component story format are still a bit 🤯  to me, despite reading https://storybook.js.org/docs/react/essentials/controls and https://storybook.js.org/docs/react/api/csf . Thus, there might be things here I've simply misunderstood.

Another unclear part is - how do we want to structure the control examples? One  (or a couple) of large examples with multiple different controls? Or one example per control, e.g.

```
Controls
  Text
  Number
  Color
  Boolean
  ...
```

Also, added a couple of comments to this PR on the parts of the code which I'm unsure about.

<!--

Everybody: Please submit all PRs to the `next-6.0` branch unless they are specific to 5.3. Storybook maintainers cherry-pick bug and documentation fixes into the `master` branch as part of the release process, so you shouldn't need to worry about this.

Maintainers: Please tag your pull request with at least one of the following:
`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

-->

dannyhw committed:
fix: eslint problems, also hooks context thing

dannyhw committed:
feat(controls): re-enable color control

lauriharpf committed:
feat: Re-enable the Boolean control (change it from a knob to a control)

lauriharpf commented on addons/ondevice-controls/src/types/Boolean.tsx:
```diff
@@ -3,40 +3,24 @@ import React from 'react';
-    );
-  }
-}
+const BooleanType = ({ arg, onChange = (value) => value }: BooleanProps) => (
```
Not sure about `= (value) => value`; is it necessary here? The typing suggests we expect an `onChange` always, so does it make sense to have a default value for it?

This situation is similar to https://github.com/storybookjs/react-native/pull/182/commits/a33b28179d3d2e23e6b26c0864cdd1ba4bc51d27#diff-7adb84faada991553c3c1a297f044b86d5b8abd6b1db5e09be8d16e66c91321cR31 , so added it just to play safe. But not sure 🤔 .

dannyhw replied:
Yeah right now I've mostly just been directly refactoring the knobs components however I think that the onchange default is not strictly necessary. To me its fine if you leave it or remove it so up to you. I expect that there will be some iterative improvements to controls. Currently it's more of a refactor of knobs but that will likely evolve.

lauriharpf replied:
Thanks! Took it out, but let's restore it if we spot a reason to have it. Not a big deal either way, was just curious 🙂 .

lauriharpf commented on addons/ondevice-controls/src/types/Boolean.tsx:
```diff
@@ -3,40 +3,24 @@ import React from 'react';
-}
+const BooleanType = ({ arg, onChange = (value) => value }: BooleanProps) => (
+  <View style={styles.switch}>
+    <Switch testID={arg.name} onValueChange={() => onChange(!arg.value)} value={arg.value} />
```
Do we have any use for the `testID` - `arg.name` -combo? Or should we just remove both? Simply carried them over from the old implementation, but if we're not using them for anything then it would make sense to just get rid of them.

dannyhw replied:
Well I think having a testID might be useful for people wanting to do automated testing on top of storybook. However I'm not sure if anyone does use it for that.

To me it's fine if you remove it since I expect to add unit tests later on which will introduce testID's in a more structured way.

lauriharpf replied:
Good point: It's possible that someone is using this. Again, not a big deal either way, but keeping it as-is might have some value. Left it for now, let's reconsider as appropriate.

dannyhw replied:
sounds good.

lauriharpf commented on addons/ondevice-controls/src/types/Boolean.tsx:
```diff
@@ -3,40 +3,24 @@ import React from 'react';
   switch: { margin: 10, alignItems: 'flex-start' },
 });
 
+BooleanType.serialize = (value) => (value ? String(value) : null);
```
I'm unsure about what these are for and why. Clearly serialization and deserialization, but why do we need them - or do we need them?

dannyhw replied:
Honestly I don't know what these are for yet, I think it might have something to do with the storybook/react-native-server package however I need to confirm that. For now I'd prefer to keep them just until I can figure out if they are needed or not.

lauriharpf replied:
Thanks - let's keep them for now. It would be interesting to know the purpose, so if you spot it (or notice that they aren't necessary) then please ping.

dannyhw replied:
Ok will let you know if I figure it out.

lauriharpf commented on examples/native/components/CSFExample/BooleanExampleButton.tsx:
```diff
@@ -0,0 +1,17 @@
+  button: { paddingHorizontal: 16, paddingVertical: 8 },
+});
+
+export const BooleanExampleButton: FunctionComponent<Props> = ({ isEnabled }: Props) => (
```
See the PR description - how do we want to structure the examples?

dannyhw replied:
This is something I'm also not sure about, previously there was one component that had examples of each control. However I do also like the idea of having stories for each one. Interested to know what your opinion is too.

lauriharpf replied:
Tough one 🤔 . Trying to list some arguments for both alternatives:

**One component with multiple controls**
+ Faster to check all control functionality (no need to jump between stories)
+ Less code

**One story per control**
+ Explicit about what code is necessary to use this control
+ Could be clearer to use as e.g. documentation examples ("see reference here")
+ Perhaps more maintainable (less complexity?). Code duplication could be tackled via shared structures if necessary.

Leaning a bit towards "one story per control", mostly because it feels like a logical structure - very explicit about what purpose each example serves. Either way, your plan (first functionality with some examples, then nicer examples) sounds reasonable.

dannyhw replied:
I agree with you, I think the one per control approach is appealing. I like the idea of it being like documentation. I think we should go down this route.

Additionally I think there is a lot that can be done with the example and it would be great to keep adding to it and improving it. I think there is already an issue on the project board too.

dannyhw replied:
Lets move this example into a ControlExamples/ folder which would contain all the control stories etc

lauriharpf committed:
refactor: Remove (probably) unnecessary onChange default from Boolean.tsx

dannyhw commented:
I just rebased the controls branch so I think I messed with the diff here. I'll rebase this branch with the controls branch to fix that.

dannyhw commented:
Had a bit of an issue with the rebase so hopefully I didn't mess anything up. Hope you don't mind me messing with your code a bit, without this rebase it was showing like 100+ changes 😅 . I think it's back to normal now though you might need to reset your local version of the branch that's all.

dannyhw reviewed (commented):
Only one thing I would change is to move the BooleanExample.stories.tsx to a different folder like ControlExamples/Boolean.stories.tsx. I need to update the other examples too, but up until this point they've mostly just been tests to see if things are working 😅 .

I'm thinking that going forward it will make sense to have this "ControlExamples" folder which would have all the controls examples.

dannyhw commented on examples/native/components/CSFExample/BooleanExample.stories.tsx:
```diff
@@ -0,0 +1,18 @@
+
+export default BooleanExample;
+
+type BooleanExampleStory = ComponentStory<typeof BooleanExampleButton>;
```
Just a minor thing, I think we can remove example from the filename and component name. Soon I will update all the others too.

lauriharpf replied:
Sounds good, did that :+1:

dannyhw reviewed (changes requested):
Like I mentioned in the comments I think it would be good to change the name and move the file into a new ControlExamples folder

lauriharpf commented:
> Had a bit of an issue with the rebase so hopefully I didn't mess anything up. Hope you don't mind me messing with your code a bit, without this rebase it was showing like 100+ changes 😅 . I think it's back to normal now though you might need to reset your local version of the branch that's all.

Nah, I don't mind at all - collective code ownership and so on 🙂 . Single-person branches are sometimes useful to avoid wasted work & for reducing the need to sync, but it's healthy to try to avoid personifying code. Let's do the best we can together 🙂 .

Good suggestions here! Will pull the branch, change these and push. The Select control turned out more complex than I initially thought - might push some initial drafts to a draft PR along with thoughts today, but probably nothing near complete. If there will be something that you can use in the draft, great. If not and you simply implement it, then that's great too - progress is what matters 😄 .

lauriharpf committed:
fix: Remove a couple of things accidentally included in rebase

dannyhw commented:
> > Had a bit of an issue with the rebase so hopefully I didn't mess anything up. Hope you don't mind me messing with your code a bit, without this rebase it was showing like 100+ changes 😅 . I think it's back to normal now though you might need to reset your local version of the branch that's all.
> 
> Nah, I don't mind at all - collective code ownership and so on 🙂 . Single-person branches are sometimes useful to avoid wasted work & for reducing the need to sync, but it's healthy to try to avoid personifying code. Let's do the best we can together 🙂 .

Great, I just wanted to check :).

> 
> Good suggestions here! Will pull the branch, change these and push. The Select control turned out more complex than I initially thought - might push some initial drafts to a draft PR along with thoughts today, but probably nothing near complete. If there will be something that you can use in the draft, great. If not and you simply implement it, then that's great too - progress is what matters 😄 .

Sounds good, if you need any help with it let me know.

lauriharpf committed:
fix: Code review fixes:

- Move examples to ControlExamples folder
- Remove unnecessary "Example"s from the boolean stories

lauriharpf commented:
> Only one thing I would change is to move the BooleanExample.stories.tsx to a different folder like ControlExamples/Boolean.stories.tsx. I need to update the other examples too, but up until this point they've mostly just been tests to see if things are working 😅 .
> 
> I'm thinking that going forward it will make sense to have this "ControlExamples" folder which would have all the controls examples.

Agreed :+1:. Did this, but please feel free to adjust the structure as necessary 👍 .

One thing that came to mind. It would feel reasonable if the stories would line up in the Storybook like this:

```
Controls
   Text
   Number
   Color
   ...
```

What we probably do not want is

```
Text control
  Basic
Number control
  Basic
Color control
  Basic
...
```

The latter would make sense if there would be wildly different examples for each control, but that doesn't seem likely at the moment. So assuming that we want everything under "Controls", we'll probably need a `Controls.stories.tsx` that has a `default` export (instead of `Boolean.stories.tsx`, which is in this PR).

dannyhw reviewed (approved):
Looks good!
